app/src/main/java/nu/yona/app/api/manager/network/BaseImpl.java

```diff
@@ -46,42 +46,38 @@
 {
 	private final int maxStale = 60 * 60 * 24 * 28; // keep cache for 28 days.
 	//    public final String localLanguage = ;
-	private final Interceptor getInterceptor = new Interceptor()
-	{
-		@Override
-		public Response intercept(Chain chain) throws IOException
+	private final Interceptor getInterceptor = chain -> {
+		Request request = chain.request();
+		if (NetworkUtils.isOnline(YonaApplication.getAppContext()))
 		{
-			Response response = null;
-			Request request = chain.request();
-			chain.request().newBuilder().addHeader(NetworkConstant.CONTENT_TYPE, "application/json");
-			if (NetworkUtils.isOnline(YonaApplication.getAppContext()))
-			{
-				chain.request().newBuilder().addHeader("Cache-Control", "only-if-cached").build();
-			}
-			else
-			{
-				throw new UnknownHostException();
-			}
-
-			response = chain.proceed(request);
-			if (response.priorResponse() != null &&
-					response.priorResponse().code() ==
-							HttpURLConnection.HTTP_MOVED_PERM)
-			{
-
-				throw new UnknownHostException();
-			}
-			else
-			{
-				request = request.newBuilder().build();
-			}
-
-			return response.newBuilder()
-					.header("Cache-Control", "public, max-age=" + maxStale)
-					.build();
+			chain.request().newBuilder().addHeader("Cache-Control", "only-if-cached").build();
+		}
+		else
+		{
+			throw new IllegalStateException(YonaApplication.getAppContext().getString(R.string.server_not_rechable));
 		}
+		chain.request().newBuilder().addHeader(NetworkConstant.CONTENT_TYPE, "application/json");
+		Response response = verifyResponse(request, chain);
+		return response.newBuilder()
+				.header("Cache-Control", "public, max-age=" + maxStale)
+				.build();
 	};
 
+	private Response verifyResponse(Request request, Interceptor.Chain chain) throws IOException
+	{
+		Response response = chain.proceed(request);
+		if (response.priorResponse() != null && response.priorResponse().code() == HttpURLConnection.HTTP_MOVED_PERM)
+		{
+			throw new IllegalStateException(YonaApplication.getAppContext().getString(R.string.server_not_rechable));
+		}
+		else if (response.code() == HttpURLConnection.HTTP_GATEWAY_TIMEOUT)
```

Comment:
There are way more error codes that don't return JSON, so the check should start with:
```
if (Family.familyOf(response.code()) != Family.SUCCESS)
```
Then you can make an attempt to interpret the response to see whether it's JSON you understand, and if not, just throw an exception that includes the status code. Otherwise, you can take the error from the interpretation of the JSON.

Reply:
In some cases we are handling 400's in the callbacks ( for e.g user override), this will disturb that logic

Reply:
Yes, I understand. That's why I talk about "attempt to...". In the server, we're also making rest requests. You might want to look into [RestClientErrorHandler.java](https://github.com/yonadev/yona-server/blob/master/core/src/main/java/nu/yona/server/rest/RestClientErrorHandler.java) to see how we "attempt to..." there.